tests/test_router_notify.cpp

```diff
@@ -0,0 +1,219 @@
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_setsockopt (
+      router, ZMQ_ROUTER_NOTIFY, &notify_opt, sizeof (notify_opt)));
+
+    // value boundary
+    notify_opt = -1;
+    TEST_ASSERT_FAILURE_ERRNO (
+      EINVAL, zmq_setsockopt (router, ZMQ_ROUTER_NOTIFY, &notify_opt,
+                              sizeof (notify_opt)));
+
+    notify_opt = (ZMQ_NOTIFY_CONNECT | ZMQ_NOTIFY_DISCONNECT) + 1;
+    TEST_ASSERT_FAILURE_ERRNO (
+      EINVAL, zmq_setsockopt (router, ZMQ_ROUTER_NOTIFY, &notify_opt,
+                              sizeof (notify_opt)));
+
+    test_context_socket_close (router);
+
+    //check a non-router socket type
+    void *dealer = test_context_socket (ZMQ_DEALER);
+    notify_opt = ZMQ_NOTIFY_CONNECT;
+    TEST_ASSERT_FAILURE_ERRNO (
+      EINVAL, zmq_setsockopt (dealer, ZMQ_ROUTER_NOTIFY, &notify_opt,
+                              sizeof (notify_opt)));
+    test_context_socket_close (dealer);
+}
+
+
+void test_router_notify_helper (int notify_opt)
+{
+    void *router = test_context_socket (ZMQ_ROUTER);
+    int opt_more;
+    size_t opt_more_length = sizeof (opt_more);
+    int opt_events;
+    size_t opt_events_length = sizeof (opt_events);
+
+
+    // valid values
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_setsockopt (
+      router, ZMQ_ROUTER_NOTIFY, &notify_opt, sizeof (notify_opt)));
+
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_bind (router, ENDPOINT_0));
+
+    void *dealer = test_context_socket (ZMQ_DEALER);
+    const char *dealer_routing_id = "X";
+
+    TEST_ASSERT_SUCCESS_ERRNO (
+      zmq_setsockopt (dealer, ZMQ_ROUTING_ID, dealer_routing_id, 1));
+
+    // dealer connects
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_connect (dealer, ENDPOINT_0));
+
+    // connection notification msg
+    if (notify_opt & ZMQ_NOTIFY_CONNECT) {
+        // routing-id only message of the connect
+        recv_string_expect_success (router, dealer_routing_id,
+                                    0);             // 1st part: routing-id
+        recv_string_expect_success (router, "", 0); // 2nd part: empty
+        TEST_ASSERT_SUCCESS_ERRNO (
+          zmq_getsockopt (router, ZMQ_RCVMORE, &opt_more, &opt_more_length));
+        TEST_ASSERT_EQUAL (0, opt_more);
+    }
+
+    // test message from the dealer
+    send_string_expect_success (dealer, "Hello", 0);
+    recv_string_expect_success (router, dealer_routing_id, 0);
+    recv_string_expect_success (router, "Hello", 0);
+    TEST_ASSERT_SUCCESS_ERRNO (
+      zmq_getsockopt (router, ZMQ_RCVMORE, &opt_more, &opt_more_length));
+    TEST_ASSERT_EQUAL (0, opt_more);
+
+    // dealer disconnects
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_disconnect (dealer, ENDPOINT_0));
+
+    // need one more process_commands() (???)
+    msleep (SETTLE_TIME);
+    zmq_getsockopt (dealer, ZMQ_EVENTS, &opt_events, &opt_events_length);
+
+    // connection notification msg
+    if (notify_opt & ZMQ_NOTIFY_DISCONNECT) {
+        // routing-id only message of the connect
+        printf ("ehe\n");
```

Comment:
Remove this? (what does "ehe" mean anyway?)

Reply:
Uhm, yeah... it's a high quality debug event, it doesn't mean anything... :) Removed.